### DISCUSSION TITLE
I wrote a prototype axum/tower authorization layer for combining auth logic using AND/OR, thoughts on similiar features in axum-login?

### DISCUSSION THREAD

**Discussion by sjud** [maxcountryman/axum-login#30] [Ideas]
Hi, sorry for long title.
I switched my graphql api on my personal project to a rest api and wrote this for describing authorization using AND/OR logic like graphql. It's not very ergonomic, also I use BoxFuture because I couldn't quite figure out how to write the future by hand. But, I was wondering if you were interested in a similar feature for axum-login or if one already exists. I think it would be good for the ecosystem to have one broad axum auth toolset, that let users describe authorization in a varied way.
https://github.com/sjud/axum_guard_logic
Warm regards, thank you for your FOSS contribution

**Comment by maxcountryman** [comment]
Thanks for sharing--super cool!

Currently route protection is provided by `axum-login` with the `RequireAuthorizationLayer` middleware via tower.

@Czocher has extended the crate `login` method to use arbitrary role types via `login_with_role`. So today applications can define ranges of roles which are able to access a resource associated with a route.

There's absolutely room to bring something like what you developed into the crate directly. It's also possible for applications to implement any arbitrary route protection they like of course but I think there's good value in providing strong primitives to build on.

**Comment by sjud** [comment]
Okay, cool! Thanks for the warm welcome.

I haven't moved to the authorization layer of my personal project yet, but when I do I will look at axum-login and axum-guard-logic and if it's okay with you I'll probably open a pull request or do my first attempt at an RFC (I haven't done this before) attempting to extend the functionality of axum-login with axum-guard-logic. I haven't done a close reading axum-guard yet so I'm not sure what my first attempt will look like. But I'll follow up here when I do so! 

Thanks! :)

**Comment by maxcountryman** [comment]
> if it's okay with you I'll probably open a pull request or do my first attempt at an RFC

Please do. If there's anything I can do to be helpful, please let me know.

**Comment by czocher** [comment]
I had a look at your crate @sjud and I'm impressed. The only drawback I could come with is the amount of nesting in some of the cases you provide.

**Comment by sjud** [comment]
@Czocher Thanks!
Yeah there are some ergonomic issues, we already use method chaining, and you could pull out some of what's written inside of the examples to make them more readable but readability is definitely issue here. I don't know how many layers will exist practically, I'm going to dogfood the crate soon and have better insight into it hopefully.
Another problem is that the service is designed to return errors fit for my particular use case whereas a more generalized auth service should have that defined generically somehow. Possibly as additional args passed to the service/layer initializing function, or some generic parameter.
